### ISSUE TITLE
General Comments

### ISSUE THREAD

**Issue by bltzr** [mrRay/OSCQueryProposal#17]
Thanks Ray for revamping the proposal, my impression is that this new brew is much more adapted to the purpose, and I also think it's good to split the query part from the plain OSC communication.
Besides, I thought it would be good to have a section for "general comments"... :-)

**Comment by jdkoftinoff** [comment]
Hi! I just stumbled on this repository and want to chime in.  I use OSC with high channel count devices.  An example of an OSC address on one device could be: 

    /input/50/eq/10/frequency ,f 1000.0

i.e. there are 72 inputs, each with 10 eq's, and each eq has frequency, gain, and Q parameters. So there is a lot of duplication here of the meta data.

As far as I can tell with your current proposal, representing this schema would require 2160 JSON objects, when really there are only 4 different variants being used.

Perhaps there could be a mechanism split the definition of an OSC address from the meta data for that kind of OSC address, and allow numeric ranges in the OSC address?

**Comment by mrRay** [comment]
yes, sometimes the JSON blobs generated by this protocol are going to be large; i've been assuming from the beginning that tens of thousands of addresses wouldn't be terribly unusual for medium-to-large applications.  did you have a specific concern about this?

>Perhaps there could be a mechanism split the definition of an OSC address from the meta data for that kind of OSC address, and allow numeric ranges in the OSC address?

did you have a specific implementation or any ideas in mind that you could share with us?  some concrete examples demonstrating the approach you're envisioning would be very helpful to this discussion!

**Comment by jdkoftinoff** [comment]
Yes, I have a specific concern about this; At Meyer Sound we use OSC for exposing control points on some products; some products of which have more than 600 filters, some with more than twice that. All of these are in orthogonal arrays. None of these devices have a service/address discovery mechanism.

Work was done by Andrew Eales from the Wellington Institute of Technology - He presented it at AES 2012 conference: http://www.aes.org/events/133/presenters/?ID=218 and see http://www.aes.org/events/133/papers/?ID=3175 and http://www.aes.org/e-lib/browse.cfm?elib=16432

He defines a complete system using OSC for the querying of the entire address space and meta data for an OSC device.

I also had published a document in 2010 shows a mechanism for introspection of the OSC schema, available at http://grouper.ieee.org/groups/1722/1/contributions/2010/1722.1-koftinoff-AVBC-2010-04-16-v3.pdf -  See Clause 5.1.

The problem with the mechanism defined in the AVBC proposal was that I did not define that the response to a query for meta information of an address could contain an address with wildcards in it.  For instance, it would have been nicer to allow a request for 

        "/osc/limits/input/1/filter/1/gain" 

to return: 

        "/osc/limits/input/*/filter/*/gain" ",sssfsfsfss", "type", "f", "min", -30.0, "max", 30.0, "inc", 0.1, "units", "dB"

Which would means that the gains for all the filters for all the inputs in the addresses:

       "/input/*/filter/*/gain"

Takes a float value with a min of -30 dB, max of 30 dB and a recommended increment of 0.1 dB

**Comment by bltzr** [comment]
Hi !

We have been addressing this problem in the jamoma project by extending the OSC syntax with the concept of instance - the separator for it being the dot (which is OSC-compliant)

the example you mention, 

> /osc/limits/input/1/filter/1/gain

would be then expressed

/osc/limits/input.1/filter.1/gain

the benefits of such a thing are several :
- "input" and "1" are basically both describing the same entity, so it seems (at least to us) better to keep them inside of the same pair of slashes
- using a different separator for instance than for nodes allows a better discrimination between, both on the human and the machine side
(this is discussed in here: Model-View-Controller separation in Max using Jamoma. <http://jamoma.org/publications/attachments/icmcsmc2014-mvc.pdf> section 3.3)
- and concerning the problem you mention, this would allow to define structures globally for each class, modulo the number of its instances (or instances of sub-classes) and thus to thin out drastically the information-load needed to describe the architecture of a system.

I don’t know if this might be useful to the protocol, but I thought it worth mentioning…

**Comment by mrRay** [comment]
>“Yes, I have a specific concern about this; At Meyer Sound we use OSC for exposing control points on some products; some products of which have more than 600 filters, some with more than twice that. All of these are in orthogonal arrays. None of these devices have a service/address discovery mechanism.”

pardon me if i’m being dense, but what is that specific concern, exactly?  if none of the devices you just mentioned have a service/address discovery mechanism then what role do they play in your concern?  can i ask for another hint?

if i seem unconcerned by the prospect of a JSON blob describing an address space with several thousand methods it’s because the main use i see for this system would involve using it with an address space with somewhere between “thousands” and “tens of thousands” of methods (depending on the user-specified setup), and most of those methods aren’t identical and aren’t split up into nice, neat numerically-indexed arrays (the address space roughly reflects the organization of the app, so there are a lot of user-named directories and the whole address space runs pretty deep).

>“Work was done by Andrew Eales from the Wellington Institute of Technology - He presented it at AES 2012 conference:”

did he post it anywhere else, or is it only behind a paywall?

>I also had published a document in 2010 shows a mechanism for introspection of the OSC schema, available at http://grouper.ieee.org/groups/1722/1/contributions/2010/1722.1-koftinoff-AVBC-2010-04-16-v3.pdf

reading it now, thanks!
- you reference a 1.1 version of the OSC spec a number of times- if you could please link me to the official OSC 1.1 spec, i’d be very grateful.  the most up-to-date info i have on the 1.1 spec is the NIME ’09 conference paper available on opensoundcontrol.org, and that explicitly isn’t a specification.  if there’s an official 1.1 spec then i’d love to read and start working off it!
- slight tangent: is it just me, or do you think OSC would benefit from a simple dictionary-type storage object?  it could be implemented with the {} chars enclosing string/value or char/value pairs in the type-tag string in a manner similar to the [] chars used to define arrays…


>“The problem with the mechanism defined in the AVBC proposal was that I did not define that the response to a query for meta information of an address could contain an address with wildcards in it.” (emphasis mine)

this is very clever, and it does a great job of reducing the amount of data you have to send for the example you provided.  as for applying this approach to the current proposal…right now, the JSON blob describing the properties of the address space also serves to enumerate the address space, and we can’t use wildcards to accomplish this.  the proposal could certainly be altered to perform enumeration of the address space as a discrete step, and then send metadata more fully describing the address space using wildcards.  that said, i have concerns about how well this approach would work in large address spaces that don’t have neatly-organized and uniform namespaces- all the test cases i have in mind would wind up using huge, enormously-complicated wildcard strings to describe methods spread all over the address space that might share the same basic attributes (and even so, several wildcard strings may be required in some circumstances).

hmm….another way of associating the same data structure with multiple OSC methods would be to explicitly define a schema- a partial OSC address path, and potentially a group of sub-paths- and then reference that schema at multiple paths in the JSON blob while enumerating the address space.  i haven’t put a ton of thought into this, but a quick example using jeff’s device could look a bit like this, with schemas that define a single EQ and a single input, and then a root node that describes 72 inputs (the schema for an input references the schema for a single EQ).

~~~
{
	"DESCRIPTION": "root node",
	"ACCESS": 0,
	"SCHEMAS": {
		"EQSchema": {
			"DESCRIPTION": "sample EQ schema",
			"ACCESS": 0,
			"CONTENTS": {
				"freq": {
					"DESCRIPTION": "frequency method",
					"ACCESS": 2,
					"TYPE": "f",
					"RANGE": [
						[0.0, 1.0, null]
					]
				},
				"gain": {
					"DESCRIPTION": "gain method",
					"ACCESS": 2,
					"TYPE": "f",
					"RANGE": [
						[0.0, 1.0, null]
					]
				},
				"Q": {
					"DESCRIPTION": "Q method",
					"ACCESS": 2,
					"TYPE": "f",
					"RANGE": [
						[0.0, 1.0, null]
					]
				}
			}
		},
		"InputSchema": {
			"DESCRIPTION": "sample input schema- note that it uses the EQSchema!",
			"ACCESS": 0,
			"CONTENTS": {
				"1": {
					"SCHEMA": "EQSchema"
				},
				"2": {
					"SCHEMA": "EQSchema"
				},
				"3": {
					"SCHEMA": "EQSchema"
				},
				"4": {
					"SCHEMA": "EQSchema"
				},
				"5": {
					"SCHEMA": "EQSchema"
				},
				"6": {
					"SCHEMA": "EQSchema"
				},
				"7": {
					"SCHEMA": "EQSchema"
				},
				"8": {
					"SCHEMA": "EQSchema"
				},
				"9": {
					"SCHEMA": "EQSchema"
				},
				"10": {
					"SCHEMA": "EQSchema"
				}
			}
		}
	}
	"CONTENTS": {
		"1": {
			"SCHEMA": "InputSchema"
		},
		"2": {
			"SCHEMA": "InputSchema"
		},
		"3": {
			"SCHEMA": "InputSchema"
		},
		"4": {
			"SCHEMA": "InputSchema"
		},
		
		... etc etc etc, didn't feel like going out to "72" :-P
	}
}
~~~

what does everybody else think?  are there any specific concerns about the length of the initial JSON blobs?  if so, what are the feelings on these workarounds, and does anyone have any other ideas?

**Comment by theod** [comment]
an idea could be : 
if the name of a node is {a, b, c} it creates 3 nodes called "a", "b" and "c". 
if the name of a node is [1::512] it creates 512 nodes called "1", "2", "3", ..., "512".

however I guess the use of { } and [ ] are not very welcome in JSON format ...
anyway mixing this kind of syntax with the schema feature could be very powerful !

**Comment by jdkoftinoff** [comment]
> but what is that specific concern, exactly? if none of the devices you just mentioned have a service/address discovery mechanism then what role do they play in your concern? can i ask for another hint?

Well, seeing that there was no standard or suitable query mechanism for OSC at the time we did not add one.  We would like to though! And it would be really nice if there were one way that was suitable for all these uses.  After a number of OSC based query mechanisms were proposed, it was made clear that this complicates the simple nature of stateless OSC; your proposal using JSON over HTTP is a great start.  I would like to be able to use it.

> did he post it anywhere else, or is it only behind a paywall?

Not that I am aware of. But I just purchased it for $5 as an AES member.

> you reference a 1.1 version of the OSC spec a number of times- if you could please link me to the official OSC 1.1 spec, 

There is no official 1.1 OSC specification.  It was expected to exist when I wrote the AVBC document but that proposal in the NIME conference paper ( http://cnmat.berkeley.edu/node/7002 ) was not carried forward.

I really like your SCHEMAS JSON entry suggestion!

**Comment by bltzr** [comment]
> I really like your SCHEMAS JSON entry suggestion!

:+1
